File: java/com/google/copybara/profiler/Profiler.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.profiler;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.base.Ticker;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;
import javax.annotation.Nullable;

/**
 * A profiler that allows to record hierarchical time statistics of the different
 * Copybara components.
 */
public final class Profiler {

  @VisibleForTesting
  public static final String ROOT_NAME = "//copybara";
  public static final String TYPE = "type";

  private final ProfilerTask nullProfilerTask;

  /**
   * A stack of tasks to be finished. When a new thread is spawned we create a copy
   * of the stack with only the top-level element so that we can use it as a parent.
   */
  @VisibleForTesting
  final InheritableThreadLocal<Deque<Task>> taskQueue =
      new InheritableThreadLocal<Deque<Task>>() {

    /**
     * Make a copy of the parent thread stack but only keep the top level one, since a child
     * shouldn't be able to finish a parent job. We only keep the latest one in order to be
     * able to construct child elements.
     */
    @Override
    protected Deque<Task> childValue(Deque<Task> parentValue) {
      // Parent can be empty if the listener creates a thread when it receives the finish
      // event for the root task ("//copybara").
      if (stopped || parentValue.isEmpty()) {
        return null;
      }
      return createQueue(parentValue.element());
    }

    /**
     * Shouldn't be called regularly. But we might have network processes that starts
     * their own thread pool outside of Copybara.
     *
     */
    @Override
    protected Deque<Task> initialValue() {
      if (stopped) {
        return null;
      }
      return createQueue(
          new Task("//detached_thread", ticker.read()));
    }
  };
  private boolean stopped;

  private Deque<Task> createQueue(Task element) {
    ArrayDeque<Task> tasks = new ArrayDeque<>(2);
    tasks.add(element);
    return tasks;
  }

  private final Ticker ticker;
  private List<Listener> listeners = ImmutableList.of();
  private ProfilerTask rootProfilerTask;

  public Profiler(Ticker ticker) {
    this.ticker = ticker;
    this.nullProfilerTask = new ProfilerTask(/* expectedTask= */ null);
  }

  /**
   * Call this method once at the beginning of Copybara binary run.
   * @param listeners the listeners to be notified o the task events
   */
  public void init(List<Listener> listeners){
    this.listeners = listeners;
    if (listeners.isEmpty()){
      return;
    }
    Task task = new Task(ROOT_NAME, ticker.read());
    taskQueue.set(createQueue(task));
    for (Listener listener : listeners) {
      listener.taskStarted(task);
    }
    rootProfilerTask = new ProfilerTask(task);
  }

  /**
   * Call this method once at the end of the Copybara binary run.
   */
  public void stop(){
    if (listeners.isEmpty()) {
      return;
    }
    Preconditions.checkState(taskQueue.get().element().getDescription().equals(ROOT_NAME));
    rootProfilerTask.close();
    Preconditions.checkState(taskQueue.get().isEmpty());
    stopped = true;
  }

  public ImmutableMap<String, String> taskType(String type) {
    return ImmutableMap.of(TYPE, type);
  }

  /**
   * Create a new profiler task that can be closed using try-with-resources.
   *
   *  <p>The profiler tasks are reentrant. So you can stack them in multiple
   *  nested try-with-resources.
   *
   * <p>Example usage:
   *  <pre>{@code
   *     try (ProfilerTask p = profiler.start("migration")) {
   *       try (ProfilerTask p2 = profiler.start("subtack")) {
   *         // Do job
   *       }
   *     }
   * }</pre>
   *
   * @param description description of the task
   * @return a {@link AutoCloseable} task than can be closed manually or with
   * try-with-resources pattern
   */
  public ProfilerTask start(String description) {
    return start(description, ImmutableMap.of());
  }

  /**
   * Overloaded method for {@link #start(String)}, that allows adding {@code fields} to the context
   * of this task.
   */
  public ProfilerTask start(String description, ImmutableMap<String, String> fields) {
    if (stopped || listeners.isEmpty()) {
      return nullProfilerTask;
    }
    Deque<Task> tasks = taskQueue.get();
    Preconditions.checkState(!tasks.isEmpty());
    Task parent = tasks.element();
    Task child = new Task(parent.getDescription() + "/" + description, fields, ticker.read());
    tasks.push(child);
    for (Listener listener : listeners) {
      listener.taskStarted(child);
    }
    return new ProfilerTask(child);
  }

  /**
   * Record a simple task metric. The user is in charge of providing its own time.
   */
  public void simpleTask(String description, long startNanos, long endNanos) {
    if (stopped || listeners.isEmpty()) {
      return;
    }
    Deque<Task> tasks = taskQueue.get();
    Preconditions.checkState(!tasks.isEmpty());
    Task parent = tasks.element();
    Task child  = new Task(parent.getDescription() + "/" + description, startNanos);
    Task finishedChild = child.finish(endNanos);
    for (Listener listener : listeners) {
      listener.taskStarted(child);
      listener.taskFinished(finishedChild);
    }
  }

  /**
   * A profiler task that can be closed to send the finish metric.
   */
  public class ProfilerTask implements AutoCloseable {
    @Nullable
    private final Task expectedTask;

    private ProfilerTask(@Nullable Task expectedTask) {
      this.expectedTask = expectedTask;
    }

    /**
     * Close the {@code task} if its not null.
     */
    @Override
    public void close() {
      if (expectedTask != null && !stopped) {
        Task task = taskQueue.get().pop();
        if (task != this.expectedTask) {
          throw new IllegalStateException("Trying to finish a task that is different"
              + " from the registered one: " + task.getDescription() + ". Expecting: "
              + this.expectedTask.getDescription());
        }
        task = task.finish(ticker.read());
        for (Listener listener : listeners) {
          listener.taskFinished(task);
        }
      }
    }
  }
}

File: src/Copybara.Core/Profiler/Profiler.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Generic;
using System.Collections.Immutable;
using Copybara.Common;

namespace Copybara.Profiler;

/// <summary>
/// A profiler that allows to record hierarchical time statistics of the different Copybara
/// components.
/// </summary>
///
/// <remarks>
/// <para>The Java original stores its per-thread task stack in an <c>InheritableThreadLocal</c>
/// whose <c>childValue</c> hook copies only the top element into a spawned thread so a child can
/// build (but not finish) descendants of its parent's current task. .NET has no
/// <c>InheritableThreadLocal</c> with a <c>childValue</c> hook; this port uses a
/// <see cref="ThreadLocal{T}"/> whose factory seeds a fresh stack with a "//detached_thread" root,
/// mirroring the Java <c>initialValue</c> path. The common single-threaded start/stop/scope
/// behaviour is preserved exactly.</para>
/// </remarks>
public sealed class Profiler
{
    public const string RootName = "//copybara";
    public const string TypeKey = "type";

    private readonly Ticker _ticker;
    private readonly ProfilerTask _nullProfilerTask;

    /// <summary>
    /// A stack of tasks to be finished. Each thread keeps its own stack; a freshly seen thread gets
    /// a "//detached_thread" root, matching Java's <c>initialValue</c>.
    /// </summary>
    private readonly ThreadLocal<Stack<Task>?> _taskQueue;

    private volatile bool _stopped;
    private IReadOnlyList<IListener> _listeners = ImmutableArray<IListener>.Empty;
    private ProfilerTask? _rootProfilerTask;

    public Profiler(Ticker ticker)
    {
        _ticker = ticker;
        _nullProfilerTask = new ProfilerTask(this, expectedTask: null);
        _taskQueue = new ThreadLocal<Stack<Task>?>(() =>
        {
            if (_stopped)
            {
                return null;
            }
            return CreateQueue(new Task("//detached_thread", _ticker.Read()));
        });
    }

    private static Stack<Task> CreateQueue(Task element)
    {
        var tasks = new Stack<Task>(2);
        tasks.Push(element);
        return tasks;
    }

    /// <summary>
    /// Call this method once at the beginning of the Copybara binary run.
    /// </summary>
    /// <param name="listeners">the listeners to be notified of the task events</param>
    public void Init(IReadOnlyList<IListener> listeners)
    {
        _listeners = listeners;
        if (listeners.Count == 0)
        {
            return;
        }
        var task = new Task(RootName, _ticker.Read());
        _taskQueue.Value = CreateQueue(task);
        foreach (var listener in listeners)
        {
            listener.TaskStarted(task);
        }
        _rootProfilerTask = new ProfilerTask(this, task);
    }

    /// <summary>
    /// Call this method once at the end of the Copybara binary run.
    /// </summary>
    public void Stop()
    {
        if (_listeners.Count == 0)
        {
            return;
        }
        var queue = _taskQueue.Value!;
        Preconditions.CheckState(queue.Peek().Description.Equals(RootName));
        _rootProfilerTask!.Close();
        Preconditions.CheckState(queue.Count == 0);
        _stopped = true;
    }

    public ImmutableDictionary<string, string> TaskType(string type) =>
        ImmutableDictionary<string, string>.Empty.Add(TypeKey, type);

    /// <summary>
    /// Create a new profiler task that can be closed using a <c>using</c> statement.
    ///
    /// <para>The profiler tasks are reentrant. So you can stack them in multiple nested
    /// <c>using</c> blocks.</para>
    ///
    /// <para>Example usage:</para>
    /// <code>
    ///     using (var p = profiler.Start("migration"))
    ///     {
    ///         using (var p2 = profiler.Start("subtask"))
    ///         {
    ///             // Do job
    ///         }
    ///     }
    /// </code>
    /// </summary>
    /// <param name="description">description of the task</param>
    /// <returns>an <see cref="IDisposable"/> task that can be closed manually or with a
    /// <c>using</c> statement</returns>
    public ProfilerTask Start(string description) =>
        Start(description, ImmutableDictionary<string, string>.Empty);

    /// <summary>
    /// Overloaded method for <see cref="Start(string)"/>, that allows adding <paramref name="fields"/>
    /// to the context of this task.
    /// </summary>
    public ProfilerTask Start(string description, ImmutableDictionary<string, string> fields)
    {
        if (_stopped || _listeners.Count == 0)
        {
            return _nullProfilerTask;
        }
        var tasks = _taskQueue.Value!;
        Preconditions.CheckState(tasks.Count != 0);
        var parent = tasks.Peek();
        var child = new Task(parent.Description + "/" + description, fields, _ticker.Read());
        tasks.Push(child);
        foreach (var listener in _listeners)
        {
            listener.TaskStarted(child);
        }
        return new ProfilerTask(this, child);
    }

    /// <summary>
    /// Record a simple task metric. The user is in charge of providing its own time.
    /// </summary>
    public void SimpleTask(string description, long startNanos, long endNanos)
    {
        if (_stopped || _listeners.Count == 0)
        {
            return;
        }
        var tasks = _taskQueue.Value!;
        Preconditions.CheckState(tasks.Count != 0);
        var parent = tasks.Peek();
        var child = new Task(parent.Description + "/" + description, startNanos);
        var finishedChild = child.Finish(endNanos);
        foreach (var listener in _listeners)
        {
            listener.TaskStarted(child);
            listener.TaskFinished(finishedChild);
        }
    }

    /// <summary>
    /// A profiler task that can be closed to send the finish metric.
    /// </summary>
    public sealed class ProfilerTask : IDisposable
    {
        private readonly Profiler _profiler;
        private readonly Task? _expectedTask;

        internal ProfilerTask(Profiler profiler, Task? expectedTask)
        {
            _profiler = profiler;
            _expectedTask = expectedTask;
        }

        /// <summary>Close the task if it's not null.</summary>
        public void Close()
        {
            if (_expectedTask != null && !_profiler._stopped)
            {
                var task = _profiler._taskQueue.Value!.Pop();
                if (task != _expectedTask)
                {
                    throw new InvalidOperationException(
                        "Trying to finish a task that is different from the registered one: "
                            + task.Description + ". Expecting: " + _expectedTask.Description);
                }
                task = task.Finish(_profiler._ticker.Read());
                foreach (var listener in _profiler._listeners)
                {
                    listener.TaskFinished(task);
                }
            }
        }

        public void Dispose() => Close();
    }
}
